src/plugins/cgal/obj.cpp

```diff
@@ -0,0 +1,267 @@
+#include "obj.h"
+
+#include <array>
+#include <cassert>
+#include <fstream>
+#include <sstream>
+#include <string>
+
+#include <boost/filesystem.hpp>
+
+#include "builder.h"
+#include "obj.h"
+
+namespace possumwood {
+
+namespace {
+struct FaceIndex {
+	int v = -1;
+	int vn = -1;
+	int vt = -1;
+};
+
+struct Face {
+	std::vector<FaceIndex> indices;
+	std::size_t objectId;
+};
+
+class FaceAdaptor {
+  public:
+	class Vertices {
+	  private:
+		static int convert(const FaceIndex& fi) {
+			return fi.v - 1;
+		}
+
+	  public:
+		Vertices(const Face* face) : m_face(face) {
+			assert(face != nullptr);
+		}
+
+		typedef boost::
+		    transform_iterator<std::function<int(const FaceIndex&)>, std::vector<FaceIndex>::const_iterator, int, int>
+		        const_iterator;
+
+		const_iterator begin() const {
+			return const_iterator(m_face->indices.begin(), convert);
+		}
+
+		const_iterator end() const {
+			return const_iterator(m_face->indices.end(), convert);
+		}
+
+		const Face* m_face;
+	};
+
+  private:
+	static const Vertices convert(const Face& face) {
+		return Vertices(&face);
+	}
+
+  public:
+	FaceAdaptor(const std::vector<Face>* faces) : m_faces(faces) {
```

Comment:
Codacy found an issue: [Class 'FaceAdaptor' has a constructor with 1 argument that is not explicit.](https://app.codacy.com/manual/martin.prazak/possumwood/pullRequest?prid=6280524)